### ISSUE TITLE
Accumulate sample 

### ISSUE THREAD

**Issue by CuongNN218** [bamps53/SeesawLoss#1]
Hi, thanks for your great work.
I saw your implementation of Seesaw Loss but i'm really curious about the accumulate of sample during training process of Mitigration factor. Did it include in your code ?

**Comment by bamps53** [comment]
Yes, this one is the original implementation actally.
https://github.com/bamps53/SeesawLoss/blob/476d60f6c8528f26491a34e397396b2b06ed380d/seesaw_loss.py#L48
Sorry for the confusing class name, I was miunderstood at first so I named SeesawLoss to the one which requires class distribution as argument.

**Comment by CuongNN218** [comment]
thanks for your quick reply, I'm really confusing about class_count. In your implementation, self.class_count is always assign None in init func which means that this line below is alway executed so you don't need else. Am I right? Please correct me if i have misunderstood.
https://github.com/bamps53/SeesawLoss/blob/476d60f6c8528f26491a34e397396b2b06ed380d/seesaw_loss.py#L68

**Comment by CuongNN218** [comment]
And one more question, what about compensation factor in Seesaw Loss, i didn't see it in your implementation.

**Comment by bamps53** [comment]
>  so you don't need else
No, this condition is true only for the very first batch, from 2nd batch class_count is accumulated one.

> what about compensation factor in Seesaw Loss, i didn't see it in your implementation.
I didn't know that what compensation factor is, but when I check the paper and it makes sense.
I implemented this based on [their technical report at LVIS workshop](https://arxiv.org/pdf/2008.10032v1.pdf), but later the arxiv link was updated!
The answer for your question is no, it's not included for now. Maybe I will read updated paper and implement:)

**Comment by CuongNN218** [comment]
https://arxiv.org/pdf/2008.10032.pdf think you should take a look in their paper for more detail on this factor, they publish this paper recently.

**Comment by CuongNN218** [comment]
@bamps53  following your implementation, i just add compensation factor to your loss, but it don't work as i expect. And then i replace my implementation with yours, the loss value go down normally. I tested on cifar 10 dataset (maybe it isn't a long-tail distribution dataset). My implementation here
`class SeesawLossWithLogits(nn.Module):
    def __init__(self, p: float = 0.8):
        super().__init__()
        self.eps = 1.0e-6
        self.p = p
        self.s = None
        self.class_counts = None

    def forward(self, logits, targets):
        
        if self.class_counts is None:
            self.class_counts = targets.sum(axis=0) + 1 # to prevent devided by zero.
        else:
            self.class_counts += targets.sum(axis=0)

        m_conditions = self.class_counts[:, None] > self.class_counts[None, :]
        m_trues = (self.class_counts[None, :] / self.class_counts[:, None]) ** self.p
        m_falses = torch.ones(len(self.class_counts), len(self.class_counts))
        m = torch.where(m_conditions, m_trues, m_falses)
        
        # calculate probs of predictions 
        sigma = F.softmax(logits, dim=1) # sigma N x C N: batch size, C: num_class

        c_conditions = sigma[:, :, None] > sigma[:, None, :]
        c_trues = (sigma[:, :, None] / sigma[:, None, :]) ** self.p
        c_falses = torch.ones(logits.size(0), len(self.class_counts), len(self.class_counts))
        c = torch.where(c_conditions, c_trues, c_falses)

        s = m * c # Sij = Mij * Cij S: B x C x C

        numerator = torch.exp(logits)
        denominator = (
            (1 - targets)[:, None, :] * s
            * torch.exp(logits)[:, None, :]).sum(axis=-1) \
            + torch.exp(logits)

        sigma = numerator / (denominator + self.eps)
        loss = (- targets * torch.log(sigma + self.eps)).sum(-1)
        return loss.mean()`

**Comment by Fan9** [comment]
@CuongNN218  I find some problems in your implementation: Only in the wrong sample, the compensation factor will be calculated.not every sample.
I tried to add compensation factor to  bamps53's loss.My implementation [here](https://github.com/Fan9/SeesawLoss_pytorch/blob/main/seesawloss.py), the loss can go down well.
I tested this loss on  a long-tail distribution dataset with K-fold cross validation(NLP classification task).The result is as follows：
ce<seesaw loss< focal loss.
i think it is possible that Normalized Linear Activation is not uesd to cause this result.
If there is any problem with my implementation, please let me know. thanks ~
